### PR TITLE
Allow omitting std::complex instantiations

### PR DESCRIPTION
**Pull request by masterleinad**
This PR aims to allow omitting instantiating all the classes for `std::complex`.
These overloads are rarely used and already cause problems (#6856, #7180).

TODO:
- [x] Disable more instantiations in the `*.cc` source files.
- [x] Restrict tests
- [x] Compatibility check with `PETSc`

### CODE REVIEW

**Comment by drwells** [comment]
I don't have the numbers in front of me right now, but adding these instantiations massively increased our compilation time and library sizes, so (for those of us who do not use them regularly) adding an option to disable them would help a lot.

Should we disable the LAPACK complex wrappers (i.e., functions starting with `c` and `z`) in this case as well?

**Comment by drwells** [comment]
Clarification: generating extra function prototypes shouldn't hurt, but we may need to guard some of the function calls in `LAPACKFullMatrix` that dispatch to different types, I think (e.g., `geev_helper`).

File: cmake/setup_cached_variables.cmake [hunk 1]
```diff
@@ -370,6 +371,39 @@ OPTION(DEAL_II_WITH_64BIT_INDICES
   )
 LIST(APPEND DEAL_II_FEATURES 64BIT_INDICES)
 
+OPTION(DEAL_II_WITH_COMPLEX_VALUES
+  "If set to OFF, the classes that take a number type are not explicitly instantiated for std::complex<float> and std::complex<double>. If PETSc is built with complex scalar type, this option must be ON."
```
Comment (tjhei) [inline cmake/setup_cached_variables.cmake]: Can we add to say that this effectively disables the support for computing with complex values?

File: source/lac/full_matrix.cc [hunk 2]
```diff
@@ -21,6 +21,11 @@ DEAL_II_NAMESPACE_OPEN
 
 #include "full_matrix.inst"
 
+#ifndef DEAL_II_WITH_COMPLEX_VALUES
+// instantiate for std::complex<double> because we use it internally.
```
Comment (tjhei) [inline source/lac/full_matrix.cc]: where is this used, maybe include this in the comment?

File: cmake/setup_cached_variables.cmake [hunk 3]
```diff
@@ -370,6 +371,39 @@ OPTION(DEAL_II_WITH_64BIT_INDICES
...
+ 
+  SET(DEAL_II_EXPAND_COMPLEX_LA_PARALLEL_VECTORS
+      "LinearAlgebra::distributed::Vector<std::complex<double> >"
+      "LinearAlgebra::distributed::Vector<std::complex<float> >"
+     )
+  SET(DEAL_II_EXPAND_COMPLEX_LA_PARALLEL_BLOCK_VECTORS
+      "LinearAlgebra::distributed::BlockVector<std::complex<double> >"
+      "LinearAlgebra::distributed::BlockVector<std::complex<float> >"
+     )
+ENDIF()
```
Comment (tamiko) [inline cmake/setup_cached_variables.cmake]: Lines 379 - 405 clutter up this file significantly.

May I ask you to refactor this statement into its own file `cmake/configure/configure_complex_values.cmake`?
(No need to create any macro logic there.)

**Comment by masterleinad** [comment]
> Would it make sense to disable this option by default? Otherwise it won't help with "problems", right?

I would love to but fear to break people's code using the default configuration. 
What do others think?

Dealing with these "problems" can hopefully be done with configure compatibility checks.

**Comment by bangerth** [comment]
On 09/26/2018 01:58 PM, Daniel Arndt wrote:
> I would love to but fear to break people's code using the default 
> configuration.
> What do others think?

While compile time is an issue for developers, the majority of people do 
it once and don't care all that much. The same is true for disk space -- 
it doesn't matter all that much any more today. So my inclination is to 
keep the default as it currently is, and only offer the option to work 
around problems with dependencies.

**Comment by masterleinad** [comment]
I did what I wanted to achieve in this PR.

File: cmake/configure/configure_3_petsc.cmake [hunk 4]
```diff
@@ -94,6 +94,25 @@ MACRO(FEATURE_PETSC_FIND_EXTERNAL var)
       SET(${var} FALSE)
     ENDIF()
 
+    # If PETSc is compiled with complex scalar type we need to have support
+    # for complex values within deal.II as well.
+    #
+    IF( PETSC_WITH_COMPLEX AND NOT DEAL_II_WITH_COMPLEX_VALUES )
+        MESSAGE(STATUS "The PETSc configuration is incompatible with the deal.II configuration: "
+        "PETSc is compiled with complex scalar type. "
+        "This requites support for complex values in deal.II as well."
```
Comment (tamiko) [inline cmake/configure/configure_3_petsc.cmake]: *requires

**Comment by tjhei** [comment]
> So my inclination is to keep the default as it currently is, and only offer the option to work around problems with dependencies.

The problem is that this then makes development and testing (including CI) more difficult because it is easy to break one or the other option without testing both. What do we gain exactly?

**Comment by masterleinad** [comment]
> With this patch, or with leaving the default as is?

This patch is (currently) leaving the default as is.

**Comment by tjhei** [comment]
> Dealing with these "problems" can hopefully be done with configure compatibility checks.

Ahh, ok. After rereading this comment I now understand. You are basically planning to error (or silently turning off complex support?) in deal.II if for example Trilinos is built without complex support?

I am still worried that this makes testing more difficult, but I don't have a better solution...

**Comment by bangerth** [comment]
I could live with turning these instantiations off if Trilinos (or some other dependency) is not compatible. We'd have to make sure that in that case the tutorial programs that need this don't compile and give a useful error message.

**Comment by tamiko** [comment]
All, what about I modify the testsuite in the following way (with this PR in mind)
 - run all variants with `WITH_COMPLEX_VALUES=ON`
 - create one variant with all external dependencies and `WITH_COMPLEX_VALUES=OFF` (full external dependencies, full testsuite)
 - create a build test variant with `WITH_COMPLEX_VALUES=OFF` and minimal/bundled configuration (just build test).

**Comment by bangerth** [comment]
On 10/04/2018 04:16 PM, Matthias Maier wrote:
> All, what about I modify the testsuite

You're not talking about the testsuite, but the collection of 
configurations that are run periodically on 'bob', right?

**Comment by tjhei** [comment]
> All, what about I modify the testsuite in the following way

Yes. I think it would be good to have at least one build run that disables this.

**Comment by dangars** [comment]
Hi, I'm working on a HDF5 interface, see PR #7020. It has `std:complex` support. I will rebase and include these commits when this PR is in master branch. I'll use `#ifdef DEAL_II_WITH_COMPLEX_VALUES`.

What is the expected behavior for the tests? Is this correct?
* `hdf5_03.with_complex_values=on.with_hdf5=true.mpirun=4.output` is executed when DEAL_II_WITH_COMPLEX_VALUES is defined
* `hdf5_03.with_complex_values=off.with_hdf5=true.mpirun=4.output` is executed when when DEAL_II_WITH_COMPLEX_VALUES is not defined

**Comment by masterleinad** [comment]
Is there anything someone wants to be addressed in this PR that isn't already covered?

**Comment by masterleinad** [comment]
> What is the expected behavior for the tests? Is this correct?
>
> `hdf5_03.with_complex_values=on.with_hdf5=true.mpirun=4.output`
> is executed when DEAL_II_WITH_COMPLEX_VALUES is defined
>
> `hdf5_03.with_complex_values=off.with_hdf5=true.mpirun=4.output`
> is executed when when DEAL_II_WITH_COMPLEX_VALUES is not defined

Yes, precisely.

**Comment by davydden** [comment]
let's move on with this PR, unless others have further comments, I think we shall merge in later today...

**Pull request merged**